File: JavaApp/src/main/java/com/charlezz/javaapp/feature/photo/PhotoAdapter.java
package com.charlezz.javaapp.feature.photo;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.DiffUtil;

import com.charlezz.javaapp.BR;
import com.charlezz.javaapp.R;
import com.charlezz.javaapp.feature.base.BindingPagedListAdapter;
import com.charlezz.javaapp.util.CommonDataBindingComponent;

import javax.inject.Inject;

public class PhotoAdapter extends BindingPagedListAdapter<Photo, LookUpViewBindingHolder> {
    private static DiffUtil.ItemCallback<Photo> diffCallback = new DiffUtil.ItemCallback<Photo>() {
        @Override
        public boolean areItemsTheSame(@NonNull Photo oldPhoto, @NonNull Photo newPhoto) {
            return oldPhoto.hashCode()==newPhoto.hashCode();
        }

        @Override
        public boolean areContentsTheSame(@NonNull Photo oldPhoto, @NonNull Photo newPhoto) {
            return oldPhoto.equals(newPhoto);
        }
    };

    private final CommonDataBindingComponent commonDataBindingComponent;

    @Inject
    public PhotoAdapter(CommonDataBindingComponent commonDataBindingComponent) {
        super(diffCallback);
        setHasStableIds(true);
        this.commonDataBindingComponent = commonDataBindingComponent;
    }

    @NonNull
    @Override
    public LookUpViewBindingHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        return new LookUpViewBindingHolder(DataBindingUtil.inflate(LayoutInflater.from(viewGroup.getContext()), R.layout.view_photo, viewGroup, false, commonDataBindingComponent));
    }

    @Override
    public long getItemId(int position) {
        return getItem(position).hashCode();
    }

    @NonNull
    @Override
    protected int getViewModelId(int viewType) {
        return BR.item;
    }
}

File: KotlinApp/src/main/java/com/charlezz/finalarchitecture/feature/photo/PhotoAdapter.kt
package com.charlezz.finalarchitecture.feature.photo

import androidx.recyclerview.widget.DiffUtil
import com.charlezz.finalarchitecture.BR
import com.charlezz.finalarchitecture.R
import com.charlezz.finalarchitecture.databinding.ViewPhotoBinding
import com.charlezz.finalarchitecture.feature.base.BaseAdapter
import com.charlezz.finalarchitecture.feature.base.BaseViewHolder

class PhotoAdapter : BaseAdapter<Photo, ViewPhotoBinding, PhotoAdapter.PhotoViewHolder>(BR.photo, diffCallback){

    override fun getViewHolderLayoutId(): Int = R.layout.view_photo

    companion object {
        val diffCallback = object : DiffUtil.ItemCallback<Photo>() {
            override fun areContentsTheSame(oldItem: Photo, newItem: Photo)= oldItem.name == newItem.name && oldItem.path == newItem.path
            override fun areItemsTheSame(oldItem: Photo, newItem: Photo) = oldItem == newItem
        }
    }

    inner class PhotoViewHolder(binding:ViewPhotoBinding):BaseViewHolder<ViewPhotoBinding>(binding)
}